File: apps/worker/src/contract/names.ts
/** The vocabularies the parent and child must spell identically from `contract/contract.json`. */

import type { AnalysisFailureReason, CountsBasis, UnitSystem } from '@gbd/db';

export const INVOCATION = {
  module: 'worker_child',
  positionalArguments: ['runDirectory'],
  workingDirectory: 'work',
  // Everything the child's environment is allowed to hold — inherited from the parent's own
  // process (PATH, HOME, LANG, TZ) plus the one secret it needs (the OpenAI key). The PDF
  // extractor's and entree detector's keys stay out: both run in the lab, outside the shipped
  // child.
  environmentVariables: ['PATH', 'HOME', 'LANG', 'TZ', 'OPENAI_API_KEY'],
} as const;

export const EXIT_CODES = {
  wroteResult: 0,
  wroteFailure: 1,
  usageError: 2,
} as const;

export const COUNTS_BASES = ['people', 'meals'] as const satisfies readonly CountsBasis[];
export const UNIT_SYSTEMS = ['lb', 'kg'] as const satisfies readonly UnitSystem[];

/** `satisfies Record<AnalysisFailureReason, …>` makes a new database enum value a compile error
 * here until someone assigns it an owner.
 */
export const FAILURE_REASON_CLAIMANT = {
  child_crashed: 'parent',
  contract_violation: 'either',
  hard_timeout: 'parent',
  hung: 'parent',
  infrastructure: 'parent',
  unknown: 'either',
  upstream_api: 'child',
  abandoned: 'parent',
  shut_down: 'parent',
  unusable_data: 'child',
} as const satisfies Record<AnalysisFailureReason, 'parent' | 'child' | 'either'>;

export type FailureReasonClaimant = (typeof FAILURE_REASON_CLAIMANT)[AnalysisFailureReason];

export const CHILD_FAILURE_REASONS = Object.entries(FAILURE_REASON_CLAIMANT)
  .filter(([, claimant]) => claimant === 'child' || claimant === 'either')
  .map(([reason]) => reason as AnalysisFailureReason)
  .sort();

File: python/worker_child/src/worker_child/contract/names.py
"""The vocabularies the parent and child must spell identically."""

from typing import Final, Literal, get_args

MODULE: Final = "worker_child"
POSITIONAL_ARGUMENTS: Final = ("runDirectory",)
WORKING_DIRECTORY: Final = "work"
SECRET_ENVIRONMENT_VARIABLES: Final = (
    "GEMINI_API_KEY",
    "LLM_WHISPERER_API_KEY",
    "OPENAI_API_KEY",
)

EXIT_WROTE_RESULT: Final = 0
EXIT_WROTE_FAILURE: Final = 1
EXIT_USAGE_ERROR: Final = 2

CountsBasis = Literal["people", "meals"]
UnitSystem = Literal["lb", "kg"]

COUNTS_BASES: Final[tuple[CountsBasis, ...]] = get_args(CountsBasis)
UNIT_SYSTEMS: Final[tuple[UnitSystem, ...]] = get_args(UnitSystem)

# A strict subset of `analysis_failure_reason`; the rest are verdicts only the parent can reach.
ChildFailureReason = Literal["contract_violation", "unknown", "upstream_api", "unusable_data"]

CHILD_FAILURE_REASONS: Final[tuple[ChildFailureReason, ...]] = get_args(ChildFailureReason)
